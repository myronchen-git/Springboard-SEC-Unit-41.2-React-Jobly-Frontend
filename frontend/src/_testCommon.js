const companies = Object.freeze([
  Object.freeze({
    handle: 'comp1',
    name: 'Company 1',
    description: 'Company 1 description.',
    numEmployees: 12,
    logoUrl: 'logo1.png',
  }),
  Object.freeze({
    handle: 'comp2',
    name: 'Company 2',
    description: 'Company 2 description.',
    numEmployees: 123,
    logoUrl: 'logo2.png',
  }),
  Object.freeze({
    handle: 'comp3',
    name: 'Company 3',
    description: 'Company 3 description.',
    numEmployees: 3,
  }),
  Object.freeze({
    handle: 'comp4',
    name: 'Company 4',
    description: 'Company 4 description.',
    logoUrl: 'logo4.png',
  }),
]);

const jobs = Object.freeze([
  Object.freeze({
    id: '1',
    title: 'Job 1',
    salary: 1,
    equity: 0.1,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
  Object.freeze({
    id: '2',
    title: 'Job 2',
    salary: 2,
    equity: 0.2,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
]);

const comp1Jobs = [];
[jobs[0], jobs[1]].forEach((job) => {
  const condensedJob = { ...job };
  delete condensedJob.companyHandle;
  delete condensedJob.companyName;
  comp1Jobs.push(Object.freeze(condensedJob));
});

const companyDetails = Object.freeze({
  handle: 'comp1',
  name: 'Company 1',
  description: 'Company 1 description.',
  numEmployees: 12,
  logoUrl: 'logo1.png',
  jobs: Object.freeze(comp1Jobs),
});

// ==================================================

export { companies, companyDetails, jobs };
