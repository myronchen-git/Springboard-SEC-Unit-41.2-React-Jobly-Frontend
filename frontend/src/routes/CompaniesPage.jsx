import { useEffect, useState } from 'react';

import JoblyApi from '../../../api';
import CompanyCard from '../components/CompanyCard';
import SearchBar from '../components/SearchBar';

// ==================================================

/**
 * Lists all companies.  Allows the list to be filtered.
 */
function CompaniesPage() {
  const [companies, setCompanies] = useState([]);

  useEffect(() => {
    retrieveCompanies();
  }, []);

  async function retrieveCompanies(filters = {}) {
    setCompanies(await JoblyApi.getCompanies(filters));
  }

  return (
    <main className="CompaniesPage">
      <SearchBar retrieveItems={retrieveCompanies} filterName="name" />
      <ul>
        {companies.map((company) => (
          <li key={company.handle}>
            <CompanyCard company={company} />
          </li>
        ))}
      </ul>
    </main>
  );
}

// ==================================================

export default CompaniesPage;
