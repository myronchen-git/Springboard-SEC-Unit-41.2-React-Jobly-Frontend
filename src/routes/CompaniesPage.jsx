import { useEffect, useState } from 'react';

import JoblyApi from '../api.js';
import CompanyCard from '../components/CompanyCard.jsx';
import SearchBar from '../components/SearchBar.jsx';

// ==================================================

/**
 * Lists all companies.  Allows the list to be filtered.
 */
function CompaniesPage() {
  const [companies, setCompanies] = useState([]);
  const [errorMessage, setErrorMessage] = useState();

  useEffect(() => {
    retrieveCompanies();
  }, []);

  async function retrieveCompanies(filters = {}) {
    try {
      const companies = await JoblyApi.getCompanies(filters);
      setCompanies(companies);
    } catch (err) {
      setErrorMessage(err[0]);
    }
  }

  return (
    <main className="CompaniesPage">
      <SearchBar retrieveItems={retrieveCompanies} filterName="name" />
      {errorMessage ? (
        <div className="CompaniesPage__error">
          <p>{errorMessage}</p>
        </div>
      ) : (
        <ul>
          {companies.map((company) => (
            <li key={company.handle}>
              <CompanyCard company={company} />
            </li>
          ))}
        </ul>
      )}
    </main>
  );
}

// ==================================================

export default CompaniesPage;
