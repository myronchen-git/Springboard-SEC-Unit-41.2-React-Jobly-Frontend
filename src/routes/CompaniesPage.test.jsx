import { act, fireEvent, render } from '@testing-library/react';
import { useNavigate } from 'react-router-dom';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { companies } from '../_testCommon.js';
import JoblyApi from '../api.js';
import CompaniesPage from './CompaniesPage.jsx';

// ==================================================

vi.mock(import('../api.js'), () => {
  const MockJoblyApi = vi.fn();
  MockJoblyApi.getCompanies = vi.fn();

  return {
    default: MockJoblyApi,
  };
});

vi.mock('react-router-dom');

// ==================================================

describe('CompaniesPage', () => {
  const mockNavigate = vi.fn();
  useNavigate.mockReturnValue(mockNavigate);

  beforeEach(() => {
    JoblyApi.getCompanies.mockImplementation((filters) => {
      let companiesToReturn;

      if (Object.keys(filters).length) {
        // If there are filters.

        const companyName = filters.name;

        companiesToReturn = companies.reduce((arr, company) => {
          if (company.name.toLowerCase().includes(companyName.toLowerCase())) {
            arr.push({ ...company });
          }

          return arr;
        }, []);
      } else {
        // If there are no filters.

        companiesToReturn = structuredClone(companies);
      }

      return Promise.resolve(companiesToReturn);
    });

    mockNavigate.mockReset();
  });

  it('renders.', async () => {
    await act(async () => render(<CompaniesPage />));
  });

  it('matches snapshot.', async () => {
    const { asFragment } = await act(async () => render(<CompaniesPage />));
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays companies and search bar.', async () => {
    // Act
    const { getByPlaceholderText, getByText } = await act(async () =>
      render(<CompaniesPage />)
    );

    // Assert
    companies.forEach((company) => {
      expect(getByText(company.name)).toBeVisible();
    });
    expect(getByPlaceholderText('Enter search term...')).toBeVisible();
  });

  it('filters the list of companies.', async () => {
    // Arrange
    const { getByPlaceholderText, getByText, queryByText } = await act(
      async () => render(<CompaniesPage />)
    );

    const searchInput = getByPlaceholderText('Enter search term...');
    const submitBtn = getByText('Submit');

    // Act
    fireEvent.change(searchInput, { target: { value: companies[0].name } });
    await act(async () => fireEvent.click(submitBtn));

    // Assert
    expect(getByText(companies[0].description)).toBeVisible();

    companies.forEach((company, i) => {
      if (i !== 0) {
        expect(queryByText(company.name)).not.toBeInTheDocument();
      }
    });
  });

  it("navigates to a specific company's page.", async () => {
    // Arrange
    const { getByText } = await act(async () => render(<CompaniesPage />));

    const companyToVisit = getByText(companies[0].name);

    // Act
    fireEvent.click(companyToVisit);

    // Assert
    expect(mockNavigate).toHaveBeenCalledOnce();
    expect(mockNavigate).toHaveBeenCalledWith(
      `/companies/${companies[0].handle}`
    );
  });
});
