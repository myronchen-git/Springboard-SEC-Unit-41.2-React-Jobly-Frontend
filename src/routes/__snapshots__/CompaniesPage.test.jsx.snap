// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CompaniesPage > matches snapshot. 1`] = `
<DocumentFragment>
  <main
    class="CompaniesPage"
  >
    bound HTMLFormElement {
      "0": <input
        name="name"
        placeholder="Enter search term..."
        type="search"
        value=""
      />,
      "1": <button
        type="submit"
      >
        Submit
      </button>,
    }
    <ul>
      <li>
        <article
          class="CompanyCard"
        >
          <header
            class="CompanyCard__header"
          >
            <img
              alt="Company 1 Logo"
              src="logo1.png"
            />
            <h1>
              Company 1
            </h1>
          </header>
          <summary
            class="CompanyCard__info"
          >
            <p>
              12 employees
            </p>
            <p>
              Company 1 description.
            </p>
          </summary>
        </article>
      </li>
      <li>
        <article
          class="CompanyCard"
        >
          <header
            class="CompanyCard__header"
          >
            <img
              alt="Company 2 Logo"
              src="logo2.png"
            />
            <h1>
              Company 2
            </h1>
          </header>
          <summary
            class="CompanyCard__info"
          >
            <p>
              123 employees
            </p>
            <p>
              Company 2 description.
            </p>
          </summary>
        </article>
      </li>
      <li>
        <article
          class="CompanyCard"
        >
          <header
            class="CompanyCard__header"
          >
            <img
              alt="Company 3 Logo"
            />
            <h1>
              Company 3
            </h1>
          </header>
          <summary
            class="CompanyCard__info"
          >
            <p>
              3 employees
            </p>
            <p>
              Company 3 description.
            </p>
          </summary>
        </article>
      </li>
      <li>
        <article
          class="CompanyCard"
        >
          <header
            class="CompanyCard__header"
          >
            <img
              alt="Company 4 Logo"
              src="logo4.png"
            />
            <h1>
              Company 4
            </h1>
          </header>
          <summary
            class="CompanyCard__info"
          >
            <p>
              Company 4 description.
            </p>
          </summary>
        </article>
      </li>
    </ul>
  </main>
</DocumentFragment>
`;
